pkg/spoke/addon/lease_controller.go

```diff
@@ -0,0 +1,191 @@
+	clusterName string
+	addOnClient addonclient.Interface
+	addOnLister addonlisterv1alpha1.ManagedClusterAddOnLister
+	leaseLister coordlisters.LeaseLister
+}
+
+// NewManagedClusterAddOnLeaseController returns an instance of managedClusterAddOnLeaseController
+func NewManagedClusterAddOnLeaseController(clusterName string,
+	addOnClient addonclient.Interface,
+	addOnInformer addoninformerv1alpha1.ManagedClusterAddOnInformer,
+	leaseInformer coordinformers.LeaseInformer,
+	resyncInterval time.Duration,
+	recorder events.Recorder) factory.Controller {
+	c := &managedClusterAddOnLeaseController{
+		clusterName: clusterName,
+		addOnClient: addOnClient,
+		addOnLister: addOnInformer.Lister(),
+		leaseLister: leaseInformer.Lister(),
+	}
+	return factory.New().
+		WithInformersQueueKeyFunc(c.queueKeyFunc, leaseInformer.Informer()).
+		WithSync(c.sync).
+		ResyncEvery(resyncInterval).
+		ToController("ManagedClusterAddOnLeaseController", recorder)
+}
+
+func (c *managedClusterAddOnLeaseController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	queueKey := syncCtx.QueueKey()
+	if queueKey == factory.DefaultQueueKey {
+		addOns, err := c.addOnLister.ManagedClusterAddOns(c.clusterName).List(labels.Everything())
+		if err != nil {
+			return err
+		}
+		errs := []error{}
+		for _, addOn := range addOns {
+			addOnConfig, err := getAddOnConfig(addOn)
+			if err != nil {
+				// failed to get addon configuration, ignore it.
+				continue
+			}
+			if err := c.syncSingle(ctx, addOnConfig.InstallationNamespace, addOn, syncCtx.Recorder()); err != nil {
+				errs = append(errs, err)
+			}
+		}
+		return operatorhelpers.NewMultiLineAggregate(errs)
+	}
+
+	addOnNamespace, addOnName, err := cache.SplitMetaNamespaceKey(queueKey)
+	if err != nil {
+		// queue key is bad format, ignore it.
+		return nil
+	}
+
+	addOn, err := c.addOnLister.ManagedClusterAddOns(c.clusterName).Get(addOnName)
+	if errors.IsNotFound(err) {
+		// addon is not found, could be deleted, ignore it.
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	return c.syncSingle(ctx, addOnNamespace, addOn, syncCtx.Recorder())
+}
+
+func (c *managedClusterAddOnLeaseController) syncSingle(ctx context.Context,
+	leaseNamespace string,
+	addOn *addonv1alpha1.ManagedClusterAddOn,
+	recorder events.Recorder) error {
+	// addon lease name should be same with the addon name.
+	observedLease, err := c.leaseLister.Leases(leaseNamespace).Get(addOn.Name)
+
+	var condition metav1.Condition
+	switch {
+	case errors.IsNotFound(err):
+		condition = metav1.Condition{
+			Type:    addOnAvailableConditionType,
+			Status:  metav1.ConditionUnknown,
+			Reason:  "ManagedClusterAddOnLeaseNotFound",
+			Message: "Managed cluster addon agent lease is not found.",
+		}
+	case err != nil:
+		return err
+	case err == nil:
+		gracePeriod := time.Duration(leaseDurationTimes*leaseDurationSeconds) * time.Second
+		if time.Now().Before(observedLease.Spec.RenewTime.Add(gracePeriod)) {
```

Comment:
I recommend a clock on `c` which you can use a `FakeClock` for testing.  I think the clock is in an apimachinery package.

Reply:
good suggestion, using `apimachinery` clock